File: pso/pso.py
import numpy as np


class Particle(object):
    """
    Particle object with position, best position and velocity properties.
    """
    def __init__(self, x, v):
        super(Particle, self).__init__()
        self.x = x  # position
        self.p = x.copy()  # best known position
        self.v = v  # velocity
        self.f_x = 0  # last function value for x
        self.f_p = 0  # last function value for p


def pso(f, swarm_size, bounds, n_iter=10, omega=0.75, phi_p=0.02, phi_g=0.1,
        verbose=False):
    """
    Particle swarm optimisation for n-dimensional spaces within bounds given.

    f: function to minimise.
    swarm_size: number of particles in swarm.
    bounds: position constraints on search space of form:
        [('name', [b_low, b_high])]
        e.g. [('x', [-1, 2]), ('y', [-2, 3])]
    n_iter: number of iterations.
    omega: previous velocity component update rate
    phi_p: particle contribution to velocity update
    phi_g: global contributio to velocity update

    """
    ps = 'running particle swarm optimisation...\n\n'
    args = (swarm_size, n_iter, omega)
    ps += 'swarm_size = %d\nn_iter = %d\nomega = %s\n' % args
    ps += 'phi_p = %s\nphi_g = %s\n' % (phi_p, phi_g)
    print ps

    n_dim = len(bounds)

    x, v = [], []  # TODO: initialise with numpy arrays
    for i, (_, b) in enumerate(bounds):
        b_low, b_high = b
        x.append(np.random.uniform(b_low, b_high, swarm_size))
        abs_b_diff = abs(b_high - b_low)
        v.append(np.random.uniform(-abs_b_diff, abs_b_diff, swarm_size))
    x, v = np.array(x), np.array(v)

    particles = [Particle(x[:, i], v[:, i]) for i in xrange(swarm_size)]

    for i, particle in enumerate(particles):
        if i == 0:
            g = particle.p.copy()
        elif f(*particle.p) < f(*g):
            g = particle.p.copy()

    f_g = f(*g)

    i = 0
    while i < n_iter:
        for j, particle in enumerate(particles):
            # update particle's velocity
            r_p, r_g = np.random.uniform(0, 1, [2, n_dim])
            particle.v = omega * particle.v
            particle.v += phi_p * r_p * (particle.p - particle.x)
            particle.v += phi_g * r_g * (g - particle.x)

            # update particle's position
            particle.x += particle.v

            # floor/cap positions to remain in bounds
            for q, (_, b) in enumerate(bounds):
                b_low, b_high = b
                if particle.x[q] > b_high:
                    particle.x[q] = b_high
                elif particle.x[q] < b_low:
                    particle.x[q] = b_low

            # update particle position if better
            f_x = f(*particle.x)
            f_p = f(*particle.p)
            if f_x < f_p:
                if verbose:
                    ps = '%d: (particle: %d) best position updated | ' % (i, j)
                    args = (particle.p, particle.x, f_p, f_g)
                    ps += 'p = %s -> x = %s [f(p) = %.6f; f(g) = %.6f]' % args
                    print ps
                particle.p = particle.x.copy()
                f_p = f(*particle.p)

                # update global position if better
                if f_p < f_g:
                    if verbose:
                        ps = '%d: new global best (particle: %d) | ' % (i, j)
                        args = (g, particle.p, f_p, f_g)
                        ps += 'g = %s -> p = %s [f(p) = %s; f(g) = %s]' % args
                        print ps

                    g = particle.p.copy()
                    ps = '%d: new global best (particle: %d) | ' % (i, j)
                    ps += 'g = %s; f(g) = %s' % (g, f_g)
                    print ps
                    f_g = f(*g)
        i += 1

    print '\nglobal best after %d iterations = %s\n\n' % (n_iter, g)
    return g


def rosenbrock(x, y):
    # The rosenbrock function: min @ (1, 1 )
    return (1 - x)**2 + 100 * (y - x**2)**2


def matyas(x, y):
    # minimum @ (0, 0)
    return 0.26 * (x**2 + y**2) - 0.48 * x * y


def polynomial(x, y, z, a=1, b=2, c=3):
    # minimum @ (a, b, c)
    return (x - a)**2 + (y - b)**2 + (z - c)**2


if __name__ == '__main__':
    np.random.seed(77)
    swarm_size = 250
    n_iter = 100

    def test_rosenbrock():
        bounds = [
            ('x', [-1, 1.5]),
            ('y', [-2, 3])]
        z = pso(f=rosenbrock, swarm_size=swarm_size, bounds=bounds,
                n_iter=n_iter, omega=0.75, phi_p=0.02, phi_g=0.1)

    def test_matyas():
        bounds = [
            ('x', [-5, 5]),
            ('y', [-5, 5])]
        z = pso(f=matyas, swarm_size=swarm_size, bounds=bounds,
                n_iter=n_iter, omega=0.75, phi_p=0.02, phi_g=0.1)

    def test_3d():
        bounds = [
            ('x', [-250, 500]),
            ('y', [-500, 50]),
            ('z', [-100, 100])]
        z = pso(f=polynomial, swarm_size=swarm_size, bounds=bounds,
                n_iter=n_iter, omega=0.75, phi_p=0.02, phi_g=0.1)

    test_rosenbrock()
    test_matyas()
    test_3d()

File: optimisation/pso.jl
using Distributions
using Plots
pyplot()


mutable struct Particle
    x::Vector{Float64}  # position
    b::Vector{Float64}  # best known position
    v::Vector{Float64}  # velocity
end
Particle(x, v) = Particle(x, x, v)


function create_particle(f::Function, search_space::AbstractArray)
    x, v = [], []
    for bound in search_space
        b_low, b_high = bound
        push!(x, rand(Uniform(bound...)))
        push!(v, rand(Uniform(-abs(b_high - b_low), abs(b_high - b_low))))
    end
    Particle(x, v)
end


function status(msg::String, i::Int64, j::Int64, from::AbstractArray,
    to::AbstractArray, f_to::Float64)
    println("$i: $j $msg | $from -> $to [f = $f_to]")
end


function pso(f::Function, swarm_size::Int64, bounds::AbstractArray,
    n_iter::Int64=50, ω::Float64=0.75, ϕp::Float64=0.02, ϕg::Float64=0.1,
    verbose::Bool=false)

    verbose && println("running particle swarm optimisation...\n\n")
    verbose && println("swarm_size = $swarm_size\nn_iter = $n_iter\nω=$ω\nϕp = $ϕp\nϕg = $ϕg\n")

    particles = [create_particle(f, bounds) for _ in 1:swarm_size]

    #TODO: there must be a more elegant way to acheive this??
    g = particles[1].b
    for (i, P) in enumerate(particles[2:end])
        if f(P.b...) < f(g...)
            g = P.b
        end
    end

    i = 0
    while i < n_iter
        for (j, P) in enumerate(particles)
            rp, rg = rand(Uniform(0, 1), 2)
            P.v = ω * P.v + (ϕp * rp * (P.b - P.x)) + (ϕg * rg * (g - P.x))
            P.x += P.v

            if f(P.x...) < f(P.b...)
                verbose && status("best position update", i, j, P.b, P.x, f(P.b...))
                P.b = P.x
                if f(P.b...) < f(g...)
                    verbose && status("new global best", i, j, g, P.x, f(P.b...))
                    g = P.b
                end
            end
        end
        i += 1
    end
    return g
end


function test_function(f::Function, bounds::AbstractArray)
    swarm_size = 200
    n_iter = 100
    ω = 0.75
    ϕp = 0.02
    ϕg = 0.1
    pso(f, swarm_size, bounds, n_iter, ω, ϕp, ϕg, false)
end


function repeated_accuracy(f::Function, search_space::AbstractArray,
    minimum_location::AbstractArray, n::Int64)
    results = []
    for i in 1:n
        a = minimum_location .- test_function(f, search_space)
        push!(results, a)
    end
    results
end


gaussian(x::Float64) = -exp(-(x - 1)^2) + 1  # min @ x=1

function rosenbrock(x::T, y::T) where T <: Float64
    (1 - x)^2 + 100(y - x^2)^2   # min @ (x, y)=(1, 1)
end

function matyas(x::T, y::T) where T <: Float64
    0.26 * (x^2 + y^2) - 0.48x * y  # min @ (x, y)=(0, 0)
end

function polynomial(x::T, y::T, z::T; a=1, b=2, c=3) where T <: Float64
    (x - a)^2 + (y - b)^2 + (z - c)^2  # minimum @ (a, b, c)
end

function rastrigin(x::T, y::T) where T <: Float64
    x^2 + y^2 - cos(18x) - cos(18y) + 2  # min @ (0, 0)
end


function test_all_functions()
    search_space = Array[[-5, 5]]
    result = test_function(gaussian, search_space)
    println("\nmin @ 1 [gaussian] | result=$result")

    search_space = Array[[-1, 1.5], [-2, 3]]
    result = test_function(rosenbrock, search_space)
    println("\nmin @ (1, 1) [rosenbrock] | result=$result")

    search_space = Array[[-5, 5], [-5, 5]]
    result = test_function(matyas, search_space)
    println("\nmin @ (0, 0) [matyas] result=$result")

    search_space = Array[[-5, 7], [-2, 14], [-9, 10]]
    result = test_function(polynomial, search_space)
    println("\nmin @ (1, 2, 3) [polynomial] result=$result")

    search_space = Array[[-1, 1], [-1, 1]]
    result = test_function(rastrigin, search_space)
    println("\nmin @ (0, 0) [rastrigin] result=$result")
end


function test_convergence()
    search_space = Array[[-1, 1], [-1, 1]]
    minimum_location = [0, 0]
    z = repeated_accuracy(rastrigin, search_space, minimum_location, 100)
    println(z)
    histogram(map(x -> norm(x), z))
end


test_all_functions()
# test_convergence()
